system/metrics/sinks/prometheus/prometheusReporterService.cpp

```diff
@@ -0,0 +1,354 @@
+{
+    IERRLOG("PrometheusReporterService encountered exception: %s", message);
+
+    HtmlPage page(HTTP_PAGE_TITLE);
+    page.appendContent(new CHtmlHeader(H1, "Internal Error"));
+
+    if (message && *message)
+        page.appendContent(new CHtmlText(message));
+    else
+        page.appendContent(new CHtmlText("Unknown Error"));
+
+    StringBuffer content;
+    page.getHtml(content);
+
+    response->setContent(content.length(), content.str());
+    response->setStatus(HTTP_STATUS_INTERNAL_SERVER_ERROR);
+
+    response->send();
+
+    return true;
+}
+
+bool PrometheusReporterService::onException(CHttpResponse * response, IException * e)
+{
+    StringBuffer errmessage;
+    if (e)
+        e->errorMessage(errmessage);
+    else
+        errmessage.set("Unknown Exception Error");
+
+    return onException(response, errmessage.str());
+}
+
+bool PrometheusReporterService::onGetNotFound(CHttpResponse * response, const char * path)
+{
+    return onMethodNotFound(response, "GET", path);
+}
+
+bool PrometheusReporterService::onPostNotFound(CHttpResponse * response, const char * path)
+{
+    return onMethodNotFound(response, "POST", path);
+}
+
+bool PrometheusReporterService::onMethodNotFound(CHttpResponse * response, const char * httpMethod, const char * path)
+{
+    HtmlPage page(HTTP_PAGE_TITLE);
+    VStringBuffer message("Target Service Not Found: '%s' '%s'", httpMethod, path);
+    page.appendContent(new CHtmlHeader(H1, message.str()));
+
+    StringBuffer content;
+    page.getHtml(content);
+
+    response->setContent(content.length(), content.str());
+    response->setStatus(HTTP_STATUS_NOT_FOUND);
+
+    response->send();
+    return true;
+}
+
+bool PrometheusReporterService::onMetricsUnavailable(CHttpRequest * request, CHttpResponse * response)
+{
+    HtmlPage page(HTTP_PAGE_TITLE);
+    VStringBuffer message("Metrics are not available: '%s'", request->queryPath());
+    page.appendContent(new CHtmlHeader(H1, message.str()));
+
+    StringBuffer content;
+    page.getHtml(content);
+
+    response->setContent(content.length(), content.str());
+    response->setStatus(HTTP_STATUS_OK);
+
+    response->send();
+    return true;
+}
+
+const char * mapMockMetricTypeToPrometheusStr(MetricType type)
+{
+    switch (type)
+    {
+        case hpccMetrics::METRICS_COUNTER:
+            return "counter";
+        case hpccMetrics::METRICS_GAUGE:
+            return "gauge";
+        default:
+            return nullptr;
+    }
+}
+
+void PrometheusReporterService::toPrometheusMetric(std::vector<std::shared_ptr<IMetric>> reportMetrics, StringBuffer & out)
+{
+    for (auto &pMetric: reportMetrics)
+    {
+        std::string name = pMetric->queryName();
+        if (!pMetric->queryDescription().empty())
+            out.append("# HELP ").append(name.c_str()).append(" ").append(pMetric->queryDescription().c_str()).append("\n");
+
+        MetricType type = pMetric->queryMetricType();
+        if (mapMockMetricTypeToPrometheusStr(type))
```

Comment:
Appears you do the conversion of the type twice. What if the conversion routine just returned something like "custom" and you reported the value as such. That way you'd always report all values. Not sure makes sense.